### PR TITLE
Tweak: Internal - Container changes for copy paste. [ED-4083]

### PR DESCRIPTION
**Pull request by iNewLegend**

### CODE REVIEW

File: assets/dev/js/editor/container/container.js [hunk 1]
```diff
@@ -419,6 +470,28 @@ export default class Container extends ArgsObject {
...
 
+	/**
+	 * Function forSomeChildrenRecursive().
+	 *
+	 * Will run over children recursively, breaks if the callback return true.
+	 *
+	 * @param {function(container:Container)} callback
+	 *
+	 */
+	forSomeChildrenRecursive( callback ) {
```
Comment (Nevoss) [inline assets/dev/js/editor/container/container.js:481]: Maybe it should return `true` or `false` like: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some

File: assets/dev/js/editor/document/elements/commands/paste.js [hunk 2]
```diff
@@ -20,106 +20,115 @@ export class Paste extends CommandHistory {
...
+
+	rebuild( containers, data, at ) {
+		// Paste at each target.
+		const result = [];
+
+		containers.forEach( ( targetContainer ) => {
+			let index = 'undefined' === typeof at ? targetContainer.view.collection.length : at;
+
+			data.forEach( ( model ) => {
+				switch ( model.elType ) {
```
Comment (Nevoss) [inline assets/dev/js/editor/document/elements/commands/paste.js:53]: Maybe extract each case into a method with an explicit name?

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/container/container.js (hunk 1 above)
Comment (iNewLegend) [reply assets/dev/js/editor/container/container.js:481]: Handled.

File: assets/dev/js/editor/document/elements/commands/paste.js (hunk 2 above)
Comment (iNewLegend) [reply assets/dev/js/editor/document/elements/commands/paste.js:53]: Out of scope, i don't want to modify the source, just make to make it extendable.

**Review by Nevoss — APPROVED** [review]

File: assets/dev/js/editor/container/container.js [hunk 3]
```diff
@@ -209,6 +209,57 @@ export default class Container extends ArgsObject {
 		return result;
 	}
 
+	/**
+	 * Function getUtilizedControls().
+	 *
+	 * Should return all controls that effecting the container.
+	 */
+	getUtilizedControls() {
```
Comment (matipojo) [inline assets/dev/js/editor/container/container.js:217]: Maybe:

```suggestion
	getCustomizedControls() {
```

File: assets/dev/js/editor/container/container.js [hunk 4]
```diff
@@ -209,6 +209,57 @@ export default class Container extends ArgsObject {
...
+			activeControls = this.settings.getActiveControls();
+
+		Object.entries( activeControls ).forEach( ( [ controlName, control ] ) => {
+			const controlValue = this.settings.get( control.name );
+
+			if ( control.global && ! controlValue?.length ) {
+				if ( this.globals.get( control.name )?.length || this.getGlobalDefault( controlName ).length ) {
+					control.global.utilized = true;
+
+					result[ controlName ] = control;
```
Comment (matipojo) [inline assets/dev/js/editor/container/container.js:228]: Check with @StyleShit how to check if it's a global value.

File: assets/dev/js/editor/container/container.js [hunk 5]
```diff
@@ -419,6 +470,30 @@ export default class Container extends ArgsObject {
...
+	 * Function forSomeChildrenRecursive().
+	 *
+	 * Will run over children recursively, breaks if the callback return true.
+	 *
+	 * @param {function(container:Container)} callback
+	 *
+	 */
+	forSomeChildrenRecursive( callback ) {
+		if ( callback( this ) ) {
+			return true;
```
Comment (matipojo) [inline assets/dev/js/editor/container/container.js:483]: What if it has children? (e.g. nested)

File: assets/dev/js/editor/container/container.js [hunk 6]
```diff
@@ -419,6 +470,30 @@ export default class Container extends ArgsObject {
...
 	}
 
+	/**
+	 * Function forSomeChildrenRecursive().
+	 *
+	 * Will run over children recursively, breaks if the callback return true.
+	 *
+	 * @param {function(container:Container)} callback
+	 *
+	 */
```
Comment (matipojo) [inline assets/dev/js/editor/container/container.js:480]: Try to find a better name

**Review by matipojo — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/container/container.js (hunk 4 above)
Comment (iNewLegend) [reply assets/dev/js/editor/container/container.js:228]: @StyleShit does its valid?

File: assets/dev/js/editor/container/container.js (hunk 3 above)
Comment (iNewLegend) [reply assets/dev/js/editor/container/container.js:217]: Why?

File: assets/dev/js/editor/container/container.js (hunk 6 above)
Comment (iNewLegend) [reply assets/dev/js/editor/container/container.js:480]: I have no idea...
@Nevoss, @StyleShit any suggestions?

File: assets/dev/js/editor/container/container.js (hunk 5 above)
Comment (iNewLegend) [reply assets/dev/js/editor/container/container.js:483]: Handled in test.

Comment (iNewLegend) [reply assets/dev/js/editor/container/container.js:483]: It should stop when callback is positive, if it's nested there are the rest of the method to handle it

File: assets/dev/js/editor/container/container.js (hunk 6 above)
Comment (StyleShit) [reply assets/dev/js/editor/container/container.js:480]: Didn't really understand what it does, but maybe `iterateChildrenRecursive()`?

Comment (iNewLegend) [reply assets/dev/js/editor/container/container.js:480]: Like in JS `array.some` will loop through array till the callback return positive result, but here it's recursive, runs over all containers in the 'children' hierarchy.

Comment (iNewLegend) [reply assets/dev/js/editor/container/container.js:480]: someChildrenRecursive @matipojo ?

Comment (StyleShit) [reply assets/dev/js/editor/container/container.js:480]: Maybe add to the function description that it's like `array.some()`.

File: assets/dev/js/editor/container/container.js (hunk 4 above)
Comment (StyleShit) [reply assets/dev/js/editor/container/container.js:228]: I already said from the beginning, `isGlobalApplied` isn't reliable,
@matipojo

Comment (matipojo) [reply assets/dev/js/editor/container/container.js:228]: How is the `isGlobalApplied` different from this check?

File: assets/dev/js/editor/container/container.js (hunk 3 above)
Comment (matipojo) [reply assets/dev/js/editor/container/container.js:217]: IF I understand you want to mark the controls that the user has been customized

File: assets/dev/js/editor/container/container.js (hunk 6 above)
Comment (matipojo) [reply assets/dev/js/editor/container/container.js:480]: Just `some`

File: assets/dev/js/editor/container/container.js (hunk 3 above)
Comment (iNewLegend) [reply assets/dev/js/editor/container/container.js:217]: No, I want the controls that effects the container, if it's a default global, so the global control is utilized.

File: assets/dev/js/editor/container/container.js (hunk 6 above)
Comment (iNewLegend) [reply assets/dev/js/editor/container/container.js:480]: Then we need to do it over all the `container.js` for the methods with the same 'logical suffix':

```
`findChildrenRecursive()` => `find()`
`forEachChildrenRecursive()` => `forEach()`
`forSomeChildrenRecursive()` => `some()`
```

@matipojo can you approve doing does in this PR?

Comment (matipojo) [reply assets/dev/js/editor/container/container.js:480]: `findChildrenRecursive()` => `recursiveFind()`...

Comment (iNewLegend) [reply assets/dev/js/editor/container/container.js]: And now?

File: assets/dev/js/editor/container/model/children-collection.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
+export default class ChildrenCollection extends Array {
```
Comment (matipojo) [inline assets/dev/js/editor/container/model/children-collection.js:1]: Make it a general util.
```suggestion
export default class RecursiveArray extends Array {
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/container/model/children-collection.js (hunk 7 above)
Comment (iNewLegend) [reply assets/dev/js/editor/container/model/children-collection.js:1]: It is not array generic the class know about the existence of `.children`.

File: assets/dev/js/editor/container/container.js (hunk 3 above)
Comment (iNewLegend) [reply assets/dev/js/editor/container/container.js]: Handled.

File: assets/dev/js/editor/container/model/children-collection.js (hunk 7 above)
Comment (iNewLegend) [reply assets/dev/js/editor/container/model/children-collection.js]: And now?

File: assets/dev/js/editor/container/container.js [hunk 8]
```diff
@@ -371,54 +423,6 @@ export default class Container extends ArgsObject {
...
-	/**
-	 * Function findChildrenRecursive().
-	 *
-	 * Will run over children recursively and pass the children to the callback till the callback returns positive value.
-	 *
-	 * @param {function(container:Container)} callback
-	 *
-	 * @returns {false|Container}
-	 */
-	findChildrenRecursive( callback ) {
```
Comment (matipojo) [inline assets/dev/js/editor/container/container.js:383]: Add BC

File: assets/dev/js/editor/container/container.js [hunk 9]
```diff
@@ -371,54 +423,6 @@ export default class Container extends ArgsObject {
...
-	/**
-	 * Function forEachChildrenRecursive().
-	 *
-	 * Will run over children recursively.
-	 *
-	 * @param {function(container:Container)} callback
-	 *
-	 * @returns {false|Container}
-	 */
-	forEachChildrenRecursive( callback ) {
```
Comment (matipojo) [inline assets/dev/js/editor/container/container.js:410]: Add BC

**Comment by iNewLegend** [comment]
ADD BC

**Review by matipojo — CHANGES_REQUESTED** [review]

**Pull request merged**